File: js/src/parser.mjs
// Precedence climbing. See docs/EXTENDING.md, "Adding an operator", step 5,
// and python/sel/parser.py, whose module docstring is the rationale.
//
// This host used to transcribe spec/grammar.md one function per production —
// parseSequence → parseList → parseAssignment → parseOr → … → parsePrimary,
// seventeen deep, with an arrow thunk at each of the nine binary helpers. That
// is a respectable style and it read as a literal rendering of the grammar, but
// it cost 35 stack frames per level of parenthesis nesting (measured here: 44
// frames at one paren, 1409 at forty, linear at 35.0), and E_DEPTH does not trip
// until 100 nested parens. Invisible on this host; fatal on Python's, whose
// default recursion limit is 1000 — which is why python/sel/parser.py was
// written in this shape first and is the reference for the rest.
//
// The payoff is not speed. It is that adding an operator stops being "a method
// in every parser, wired into the chain in the same place" and becomes a row in
// the table below.
//
// The depth arithmetic is unchanged and is not free to change: conformance/
// 10-limits.selt pins two increments per paren (parseSequence + parsePrimary),
// E_DEPTH at 1:101 for 100 parens, 1:200 for a `-` chain and 1:797 for a NOT
// chain. Prefix operators are counted only when actually consumed.

import { fail, MAX_DEPTH } from './errors.mjs';
import * as D from './decimal.mjs';
import { tokenize, RESERVED } from './lexer.mjs';
import { lookup } from './registry.mjs';


const ASSIGN_OPS = new Set(['=', '+=', '-=', '*=', '/=', '%=', '&=']);
const COMPARE_OPS = new Set([
  '==', '!=', '<', '<=', '>', '>=', '$==', '$!=', '$<', '$<=', '$>', '$>=',
]);
const COMPARE_WORDS = new Set(['EQL', 'IN']);

// spec/SPEC.md §5, as a table. Higher binds tighter. The gaps are the levels
// that are not infix: 16 is postfix/primary, 15 is unary minus, 7 is NOT.
const BP_SEQ = 1;       // ;
const BP_LIST = 2;      // ,
const BP_ASSIGN = 3;    // = += -= *= /= %= &=   (right associative)
const BP_OR = 4;
const BP_XOR = 5;
const BP_AND = 6;
const BP_NOT = 7;       // prefix
const BP_COMPARE = 8;   // non-associative
const BP_COALESCE = 9;  // ?? ??? (right associative)
const BP_BOR = 10;
const BP_BXOR = 11;
const BP_BAND = 12;
const BP_CONCAT = 13;   // &
const BP_ADD = 14;      // + -
const BP_MUL = 15;      // * / %
const BP_NEG = 16;      // prefix

// BP_SEQ and BP_LIST are deliberately unused: `;` and `,` build N-ary nodes, so
// they stay hand-written loops in parseSequence/parseList rather than table
// rows. They are declared anyway so the ladder above reads as spec/SPEC.md §5
// does, with no silent gap at the loose end.

// Operator -> [binding power, associativity]. 'L' left, 'R' right, 'N'
// non-associative. Word operators lex as identifiers and symbol operators as
// `op` tokens, so they are two tables sharing one set of binding powers.
//
// Map rather than a plain object because the key is token text: a bare `{}`
// would answer for every name on Object.prototype, and the parser would then
// have to assume no token can ever be spelled like one. Map has no such chain,
// and `.get` transcribes Python's `dict.get` directly.
const INFIX_OPS = new Map([
  ['??', [BP_COALESCE, 'R']],
  ['???', [BP_COALESCE, 'R']],
  ['&', [BP_CONCAT, 'L']],
  ['+', [BP_ADD, 'L']], ['-', [BP_ADD, 'L']],
  ['*', [BP_MUL, 'L']], ['/', [BP_MUL, 'L']], ['%', [BP_MUL, 'L']],
  ...[...ASSIGN_OPS].map((op) => [op, [BP_ASSIGN, 'R']]),
  ...[...COMPARE_OPS].map((op) => [op, [BP_COMPARE, 'N']]),
]);
const INFIX_WORDS = new Map([
  ['OR', [BP_OR, 'L']], ['XOR', [BP_XOR, 'L']], ['AND', [BP_AND, 'L']],
  ['BOR', [BP_BOR, 'L']], ['BXOR', [BP_BXOR, 'L']], ['BAND', [BP_BAND, 'L']],
  ...[...COMPARE_WORDS].map((w) => [w, [BP_COMPARE, 'N']]),
]);

class Parser {
  constructor(tokens) {
    this.toks = tokens;
    this.i = 0;
    this.depth = 0;
  }

  // The two tables are one lookup. Every question about an operator — what it
  // binds at, how it associates, and whether it may follow a comparison — is
  // answered from here, so adding an operator really is adding a row. Asking a
  // separate list anywhere would put that claim back in doubt.
  infixEntry(t) {
    if (t.type === 'op') return INFIX_OPS.get(t.value);
    if (t.type === 'ident') return INFIX_WORDS.get(t.value);
    return undefined;
  }

  peek() { return this.toks[this.i]; }
  next() { return this.toks[this.i++]; }
  atOp(v) { const t = this.peek(); return t.type === 'op' && t.value === v; }
  atEof() { return this.peek().type === 'eof'; }

  expectOp(v) {
    if (!this.atOp(v)) {
      const t = this.peek();
      fail('E_SYNTAX', `expected ${JSON.stringify(v)}, got ${describe(t)}`, t);
    }
    return this.next();
  }

  enter(pos) {
    if (++this.depth > MAX_DEPTH) fail('E_DEPTH', 'expression nested too deeply', pos);
  }
  leave() { this.depth--; }

  // --- entry ----------------------------------------------------------------

  parseProgram() {
    const node = this.parseSequence();
    if (!this.atEof()) {
      const t = this.peek();
      fail('E_SYNTAX', `unexpected ${describe(t)}`, t);
    }
    return node;
  }

  // sequence = list { ";" list } [ ";" ]
  //
  // The try/finally is new here. It costs nothing — a failing parse abandons the
  // Parser either way — and the Lisp and Python hosts already protect this
  // counter, so this is the shape the five hosts converged on rather than a
  // deviation. All five protect it now.
  parseSequence() {
    const start = this.peek();
    this.enter(start);
    let items;
    try {
      items = [this.parseList()];
      while (this.atOp(';')) {
        this.next();
        // A trailing ';' before a closer or end of input is permitted.
        if (this.atEof() || this.atOp(')') || this.atOp(']')) break;
        items.push(this.parseList());
      }
    } finally {
      this.leave();
    }
    return items.length === 1 ? items[0] : { t: 'seq', items, pos: items[0].pos };
  }

  // list = assignment { "," assignment }
  parseList() {
    const items = [this.parseTerm(BP_ASSIGN)];
    while (this.atOp(',')) {
      this.next();
      items.push(this.parseTerm(BP_ASSIGN));
    }
    return items.length === 1 ? items[0] : { t: 'list', items, pos: items[0].pos };
  }

  // --- the precedence-climbing loop -----------------------------------------

  parseTerm(minBp) {
    let left = this.parsePrefix(minBp);

    for (;;) {
      const t = this.peek();
      const entry = this.infixEntry(t);
      if (entry === undefined) return left;
      const [bp, assoc] = entry;
      if (bp < minBp) return left;

      this.next();

      if (ASSIGN_OPS.has(t.value)) {
        // Assignment. The target is validated against the AST shape, not against
        // a value, which is what makes `(A) = 1` a compile error. Parsing the
        // right side at bp rather than bp + 1 is what makes it right associative.
        checkTarget(left, t);
        // Counted, for the same reason parsePrefix counts: the right side recurses
        // without passing through parseSequence or parsePrimary, so uncounted a
        // chain of assignments is bounded by nothing but the host's own stack.
        // `A=` fifty thousand times segfaulted the C++ host through its public
        // CLI and raised a host RangeError here -- the same hole the prefix
        // operators had, through the one production nobody had counted.
        this.enter(t);
        try {
          const value = this.parseTerm(bp);
          left = { t: 'assign', op: t.value, target: left, value, pos: left.pos };
        } finally {
          this.leave();
        }
        continue;
      }

      if (assoc === 'R') {
        const right = this.parseTerm(bp);
        left = { t: 'bin', op: t.value, l: left, r: right, pos: t };
        continue;
      }

      if (assoc === 'N') {
        const right = this.parseTerm(bp + 1);
        const after = this.peek();
        const afterEntry = this.infixEntry(after);
        if (afterEntry !== undefined && afterEntry[1] === 'N') {
          fail('E_SYNTAX',
            `comparison operators do not chain — parenthesise, as in (a ${t.value} b) AND (b ${after.value} c)`,
            after);
        }
        left = { t: 'bin', op: t.value, l: left, r: right, pos: t };
        continue;
      }

      const right = this.parseTerm(bp + 1);
      left = { t: 'bin', op: t.value, l: left, r: right, pos: t };
    }
  }

  // NOT and unary minus.
  //
  // Each is accepted only where its own binding power reaches: NOT at 7 cannot
  // appear inside a comparison operand (parsed at 9), so `a == NOT b` falls
  // through to parsePrimary, which sees the bare identifier NOT and raises
  // E_RESERVED — the same error the transcribed parser gave, by a different
  // route. Folding these into parsePrimary, which is where textbook precedence
  // climbing puts prefix operators, would make `NOT a == b` parse as
  // `(NOT a) == b` and would break the depth pins at the same time.
  //
  // Counted, and only when actually consumed: a prefix operator recurses without
  // passing through parseSequence or parsePrimary, and uncounted it reached the
  // host's own stack limit instead of E_DEPTH — a RangeError here and a segfault
  // in C++, from a rule that is just `-` repeated.
  parsePrefix(minBp) {
    const t = this.peek();

    if (t.type === 'ident' && t.value === 'NOT' && minBp <= BP_NOT) {
      this.next();
      this.enter(t);
      try {
        return { t: 'un', op: 'NOT', x: this.parseTerm(BP_NOT), pos: t };
      } finally {
        this.leave();
      }
    }

    if (t.type === 'op' && t.value === '-' && minBp <= BP_NEG) {
      this.next();
      this.enter(t);
      try {
        return { t: 'un', op: 'NEG', x: this.parseTerm(BP_NEG), pos: t };
      } finally {
        this.leave();
      }
    }

    return this.parsePostfix();
  }

  // postfix = primary { "[" sequence "]" }
  //
  // The bracket counts a level of its own. Without it an index is the one
  // nesting door that recurses from outside parsePrimary's enter/leave, so it
  // charged one level per nesting where "(", "f(" and the prefix operators all
  // charge for the frames they actually cost. Five stack frames against one
  // level of the budget put a[a[...]] over CPython's 1000-frame limit before
  // the 200-level guard could fire, which is why the Python host raised
  // RecursionError at 198 while the others still answered.
  parsePostfix() {
    let node = this.parsePrimary();
    while (this.atOp('[') || this.atOp('.>')) {
      if (this.atOp('[')) {
        const br = this.next();
        this.enter(br);
        try {
          const idx = this.parseSequence();
          this.expectOp(']');
          node = { t: 'index', obj: node, idx, pos: br };
        } finally {
          this.leave();
        }
      } else {
        this.next();
        node = this.parsePipeStep(node);
      }
    }
    return node;
  }

  parsePipeStep(left) {
    const t = this.peek();
    if (t.type !== 'ident' || t.value === 'TRUE' || t.value === 'FALSE' || t.value === 'NULL') {
      fail('E_SYNTAX', 'right-hand side of .> must be a function call or function name', t);
    }
    const nameTok = this.next();
    let args = [];
    if (this.atOp('(')) {
      this.next();
      if (this.atOp(')')) {
        this.next();
      } else {
        const inner = this.parseSequence();
        this.expectOp(')');
        args = (inner.t === 'list' && !inner.grouped) ? inner.items : [inner];
      }
    }

    const spec = lookup(nameTok.value);
    if (!spec) fail('E_UNKNOWN_FUNC', `unknown function ${nameTok.value}`, nameTok);

    const count = args.length;
    let hasPlaceholder = false;
    if (count >= spec.min) {
      for (let i = 0; i < args.length; i++) {
        if (args[i].t === 'var' && args[i].name === '_' && !args[i].grouped) {
          args[i] = left;
          hasPlaceholder = true;
        }
      }
    }
    if (!hasPlaceholder) {
      args.unshift(left);
    }

    const finalCount = args.length;
    if (finalCount < spec.min || finalCount > spec.max) {
      fail('E_ARITY', `${spec.name} takes ${arityText(spec)}, got ${finalCount}`, nameTok);
    }
    if (spec.arityError) {
      const problem = spec.arityError(finalCount);
      if (problem) fail('E_ARITY', problem, nameTok);
    }
    return { t: 'call', name: spec.name, spec, args, pos: nameTok };
  }

  parsePrimary() {
    const t = this.peek();
    this.enter(t);
    try {
      if (t.type === 'num') {
        this.next();
        // Canonicalised once, here: the literal 007 is the value 7.
        return { t: 'num', v: D.format(D.parse(t.value, t)), pos: t };
      }
      if (t.type === 'text') { this.next(); return { t: 'text', v: t.value, pos: t }; }

      if (t.type === 'ident') {
        if (t.value === 'TRUE' || t.value === 'FALSE') {
          this.next();
          return { t: 'bool', v: t.value === 'TRUE', pos: t };
        }
        if (t.value === 'NULL') {
          this.next();
          return { t: 'null', pos: t };
        }
        const after = this.toks[this.i + 1];
        if (after && after.type === 'op' && after.value === '(') return this.parseCall();
        if (RESERVED.has(t.value)) {
          fail('E_RESERVED', `${t.value} is a reserved word and cannot be a variable`, t);
        }
        this.next();
        return { t: 'var', name: t.value, pos: t };
      }

      if (t.type === 'op' && t.value === '(') {
        this.next();
        if (this.atOp(')')) fail('E_SYNTAX', 'empty parentheses', t);
        const inner = this.parseSequence();
        this.expectOp(')');
        // Marked so that F((1,2)) passes one list rather than two arguments.
        inner.grouped = true;
        return inner;
      }

      fail('E_SYNTAX', `unexpected ${describe(t)}`, t);
    } finally {
      this.leave();
    }
  }

  parseCall() {
    const nameTok = this.next();
    this.expectOp('(');
    let args;
    if (this.atOp(')')) {
      this.next();
      args = [];
    } else {
      const inner = this.parseSequence();
      this.expectOp(')');
      args = (inner.t === 'list' && !inner.grouped) ? inner.items : [inner];
    }

    const spec = lookup(nameTok.value);
    if (!spec) fail('E_UNKNOWN_FUNC', `unknown function ${nameTok.value}`, nameTok);
    if (args.length < spec.min || args.length > spec.max) {
      fail('E_ARITY', `${spec.name} takes ${arityText(spec)}, got ${args.length}`, nameTok);
    }
    if (spec.arityError) {
      const problem = spec.arityError(args.length);
      if (problem) fail('E_ARITY', problem, nameTok);
    }
    return { t: 'call', name: spec.name, spec, args, pos: nameTok };
  }
}

function arityText(spec) {
  if (spec.max === Infinity) return `at least ${spec.min} argument${spec.min === 1 ? '' : 's'}`;
  if (spec.min === spec.max) return `${spec.min} argument${spec.min === 1 ? '' : 's'}`;
  return `${spec.min} to ${spec.max} arguments`;
}

function describe(t) {
  if (t.type === 'eof') return 'end of input';
  if (t.type === 'text') return 'a text literal';
  if (t.type === 'num') return `number ${t.value}`;
  return JSON.stringify(t.value);
}

// The target must be an identifier followed by zero or more index operations.
function checkTarget(node, opTok) {
  let n = node;
  while (n.t === 'index') n = n.obj;
  if (n.t !== 'var' || node.grouped) {
    fail('E_BAD_ASSIGN', `cannot assign with ${opTok.value} to this expression`, node.pos);
  }
}

export function parse(source) {
  return new Parser(tokenize(source)).parseProgram();
}

File: python/sel/parser.py
"""Parser.

**This is the precedence-climbing pilot.** The other four hosts transcribe
spec/grammar.md one method per production — `parse_sequence` -> `parse_list` ->
`parse_assignment` -> `parse_or` -> ... -> `parse_primary`, sixteen deep, plus a
thunk frame per binary helper. That shape is a deliberate choice there and it
costs those hosts nothing, but it is **35 stack frames per level of parenthesis
nesting** (measured against the JS host: 44 frames at one paren, 1409 at forty,
linear at 35.0). E_DEPTH trips at 100 nested parens, so transcribing it here
would need ~3500 Python frames against a default recursion limit of 1000, and
this host would raise RecursionError where the other four raise E_DEPTH.

Raising sys.setrecursionlimit would paper over that. Precedence climbing removes
it: the sixteen levels of spec/SPEC.md §5 become the table below, and one level
of nesting costs six frames instead of thirty-five.

The intent is that the other four hosts adopt this shape in turn, so the table
is written to be transcribed rather than to be clever, and anything specific to
Python is kept out of the loop.

Five things this has to reproduce exactly, none of which the type checker will
catch for you — they all produce a *valid parse of the wrong tree*:

 1. NOT is a LOOSE prefix operator (bp 7 — looser than comparison, tighter than
    AND) while unary `-` is a TIGHT one (bp 15). Textbook precedence climbing
    puts every prefix operator in parse_primary, at the tightest end, which
    would make `NOT a == b` parse as `(NOT a) == b`. Prefix operators get their
    own binding power here, and `parse_prefix` refuses one whose binding power
    is looser than the position allows — which is what makes `a == NOT b` and
    `-NOT x` E_RESERVED, exactly as the transcribed parsers make them.
 2. Comparison is NON-associative. Its right side is parsed one level tighter
    and a second comparison operator afterwards is E_SYNTAX, reported at that
    second operator.
 3. `,` and `;` build N-ary list/seq nodes, not left-leaning binary trees.
    dependencies() walks `items`, and parse_call flattens a top-level `list`
    into an argument vector; binary nodes would break both.
 4. The `grouped` flag marks a parenthesised node. It is what makes `F((1,2))`
    pass one list rather than two arguments, and `(A) = 1` an E_BAD_ASSIGN.
 5. Node positions: `bin` and `un` take the operator token's, `list`/`seq` take
    items[0]'s, `assign` takes the target's. Every pinned position in
    conformance/12-misuse.selt depends on these.

Depth is entered in parse_sequence and parse_primary — the same two places the
other hosts use, so the trip point stays two per paren and lim.parse-depth still
reports E_DEPTH at 1:101 — plus prefix operators, which are counted for the
reason given on parse_prefix.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from . import decimal as D
from .errors import MAX_DEPTH, Pos, fail
from .lexer import RESERVED, Token, tokenize
from .registry import Spec, lookup

ASSIGN_OPS = frozenset(['=', '+=', '-=', '*=', '/=', '%=', '&='])
COMPARE_OPS = frozenset(['==', '!=', '<', '<=', '>', '>=',
                         '$==', '$!=', '$<', '$<=', '$>', '$>='])
COMPARE_WORDS = frozenset(['EQL', 'IN'])

# spec/SPEC.md §5, as a table. Higher binds tighter. The gaps are the levels
# that are not infix: 16 is postfix/primary, 15 is unary minus, 7 is NOT.
BP_SEQ = 1        # ;
BP_LIST = 2       # ,
BP_ASSIGN = 3     # = += -= *= /= %= &=   (right associative)
BP_OR = 4
BP_XOR = 5
BP_AND = 6
BP_NOT = 7        # prefix
BP_COMPARE = 8    # non-associative
BP_BOR = 9
BP_BXOR = 10
BP_BAND = 11
BP_CONCAT = 12    # &
BP_ADD = 13       # + -
BP_MUL = 14       # * / %
BP_NEG = 15       # prefix

# Infix operator -> (binding power, associativity). 'L' left, 'R' right,
# 'N' non-associative. Word operators are lexed as identifiers, so they are
# looked up separately; the binding powers are the same table.
INFIX_OPS: dict[str, tuple[int, str]] = {
    '&': (BP_CONCAT, 'L'),
    '+': (BP_ADD, 'L'), '-': (BP_ADD, 'L'),
    '*': (BP_MUL, 'L'), '/': (BP_MUL, 'L'), '%': (BP_MUL, 'L'),
    **{op: (BP_ASSIGN, 'R') for op in ASSIGN_OPS},
    **{op: (BP_COMPARE, 'N') for op in COMPARE_OPS},
}
INFIX_WORDS: dict[str, tuple[int, str]] = {
    'OR': (BP_OR, 'L'), 'XOR': (BP_XOR, 'L'), 'AND': (BP_AND, 'L'),
    'BOR': (BP_BOR, 'L'), 'BXOR': (BP_BXOR, 'L'), 'BAND': (BP_BAND, 'L'),
    **{w: (BP_COMPARE, 'N') for w in COMPARE_WORDS},
}


@dataclass(slots=True)
class Node:
    t: str
    pos: Pos
    v: Any = None                       # num / text / bool literal
    name: str = ''                      # var, call
    op: str = ''                        # bin, un, assign
    l: Node | None = None               # bin left
    r: Node | None = None               # bin right
    x: Node | None = None               # un operand
    obj: Node | None = None             # index object
    idx: Node | None = None             # index key
    target: Node | None = None          # assign target
    value: Node | None = None           # assign value
    items: list[Node] = field(default_factory=list)     # seq, list
    args: list[Node] = field(default_factory=list)      # call
    spec: Spec | None = None            # call
    grouped: bool = False


class Parser:
    __slots__ = ('toks', 'i', 'depth')

    def __init__(self, tokens: list[Token]) -> None:
        self.toks = tokens
        self.i = 0
        self.depth = 0

    # --- token helpers --------------------------------------------------------

    def infix_entry(self, t: Token) -> tuple[int, str] | None:
        """The two tables are one lookup.

        Every question about an operator -- what it binds at, how it associates,
        and whether it may follow a comparison -- is answered from here, so
        adding an operator really is adding a row. Asking a separate list
        anywhere would put that claim back in doubt.
        """
        if t.type == 'op':
            return INFIX_OPS.get(t.value)
        if t.type == 'ident':
            return INFIX_WORDS.get(t.value)
        return None

    def peek(self) -> Token:
        return self.toks[self.i]

    def next(self) -> Token:
        t = self.toks[self.i]
        self.i += 1
        return t

    def at_op(self, v: str) -> bool:
        t = self.peek()
        return t.type == 'op' and t.value == v


    def at_eof(self) -> bool:
        return self.peek().type == 'eof'

    def expect_op(self, v: str) -> Token:
        if not self.at_op(v):
            t = self.peek()
            fail('E_SYNTAX', f'expected "{v}", got {describe(t)}', t.pos)
        return self.next()

    def enter(self, pos: Pos) -> None:
        self.depth += 1
        if self.depth > MAX_DEPTH:
            fail('E_DEPTH', 'expression nested too deeply', pos)

    def leave(self) -> None:
        self.depth -= 1

    # --- entry ----------------------------------------------------------------

    def parse_program(self) -> Node:
        node = self.parse_sequence()
        if not self.at_eof():
            t = self.peek()
            fail('E_SYNTAX', f'unexpected {describe(t)}', t.pos)
        return node

    # sequence = list { ";" list } [ ";" ]
    #
    # The try/finally around the depth counter is deliberate and differs from the
    # JS, PHP and C++ parsers, which leave parse_sequence's leave() unprotected.
    # It costs nothing — a failing parse abandons the Parser either way — and the
    # Lisp host's with-depth macro already protects it, so this is the shape the
    # other three should converge on rather than a Python deviation.
    def parse_sequence(self) -> Node:
        start = self.peek()
        self.enter(start.pos)
        try:
            items = [self.parse_list()]
            while self.at_op(';'):
                self.next()
                # A trailing ';' before a closer or end of input is permitted.
                if self.at_eof() or self.at_op(')') or self.at_op(']'):
                    break
                items.append(self.parse_list())
        finally:
            self.leave()
        if len(items) == 1:
            return items[0]
        return Node('seq', items[0].pos, items=items)

    # list = assignment { "," assignment }
    def parse_list(self) -> Node:
        items = [self.parse_term(BP_ASSIGN)]
        while self.at_op(','):
            self.next()
            items.append(self.parse_term(BP_ASSIGN))
        if len(items) == 1:
            return items[0]
        return Node('list', items[0].pos, items=items)

    # --- the precedence-climbing loop ----------------------------------------

    def parse_term(self, min_bp: int) -> Node:
        left = self.parse_prefix(min_bp)

        while True:
            t = self.peek()
            entry = self.infix_entry(t)
            if entry is None:
                return left
            bp, assoc = entry
            if bp < min_bp:
                return left

            self.next()

            if assoc == 'R':
                # Assignment. The target is validated against the AST shape, not
                # against a value, which is what makes `(A) = 1` a compile error.
                check_target(left, t)
                # Counted, for the same reason parse_prefix counts: the right
                # side recurses without passing through parse_sequence or
                # parse_primary, so uncounted a chain of assignments is bounded
                # by nothing but the host's own stack -- and this host has the
                # least of it, raising RecursionError where four others were
                # still returning E_DEPTH.
                self.enter(t.pos)
                try:
                    value = self.parse_term(bp)
                    left = Node('assign', left.pos, op=t.value, target=left, value=value)
                finally:
                    self.leave()
                continue

            if assoc == 'N':
                right = self.parse_term(bp + 1)
                after = self.peek()
                after_entry = self.infix_entry(after)
                if after_entry is not None and after_entry[1] == 'N':
                    fail('E_SYNTAX',
                         'comparison operators do not chain — parenthesise, as in '
                         f'(a {t.value} b) AND (b {after.value} c)',
                         after.pos)
                left = Node('bin', t.pos, op=t.value, l=left, r=right)
                continue

            right = self.parse_term(bp + 1)
            left = Node('bin', t.pos, op=t.value, l=left, r=right)

    def parse_prefix(self, min_bp: int) -> Node:
        """NOT and unary minus.

        Each is accepted only where its own binding power reaches: NOT at bp 7
        cannot appear inside a comparison operand (parsed at bp 9), so `a == NOT b`
        falls through to parse_primary, which sees the identifier NOT and raises
        E_RESERVED — the same error the transcribed parsers give, by a different
        route.

        Counted, for the same reason the transcribed parsers now count it: a
        prefix operator recurses without passing through parse_sequence or
        parse_primary, and uncounted it reached the host's own stack limit
        instead of E_DEPTH — a segfault in C++ and a RecursionError here.
        """
        t = self.peek()

        if t.type == 'ident' and t.value == 'NOT' and min_bp <= BP_NOT:
            self.next()
            self.enter(t.pos)
            try:
                return Node('un', t.pos, op='NOT', x=self.parse_term(BP_NOT))
            finally:
                self.leave()

        if t.type == 'op' and t.value == '-' and min_bp <= BP_NEG:
            self.next()
            self.enter(t.pos)
            try:
                return Node('un', t.pos, op='NEG', x=self.parse_term(BP_NEG))
            finally:
                self.leave()

        return self.parse_postfix()

    # postfix = primary { "[" sequence "]" }
    # postfix = primary { "[" sequence "]" }
    #
    # The bracket counts a level of its own. Without it an index is the one
    # nesting door that recurses from outside parse_primary's enter/leave, so it
    # charged one level per nesting where "(", "f(" and the prefix operators all
    # charge for the frames they actually cost. Five stack frames against one
    # level of the budget put a[a[...]] over CPython's 1000-frame limit before
    # the 200-level guard could fire, and this is the host where that showed:
    # `a[` x198 raised RecursionError through the public CLI while the others
    # still answered. Counting the bracket halves the density to 2.5 frames per
    # level, which puts the guard back in front of the stack at every depth.
    def parse_postfix(self) -> Node:
        node = self.parse_primary()
        while self.at_op('['):
            br = self.next()
            self.enter(br.pos)
            try:
                idx = self.parse_sequence()
                self.expect_op(']')
                node = Node('index', br.pos, obj=node, idx=idx)
            finally:
                self.leave()
        return node

    def parse_primary(self) -> Node:
        t = self.peek()
        self.enter(t.pos)
        try:
            if t.type == 'num':
                self.next()
                # Canonicalised once, here: the literal 007 is the value 7.
                return Node('num', t.pos, v=D.format(D.parse(t.value, t.pos)))

            if t.type == 'text':
                self.next()
                return Node('text', t.pos, v=t.value)

            if t.type == 'ident':
                if t.value in ('TRUE', 'FALSE'):
                    self.next()
                    return Node('bool', t.pos, v=(t.value == 'TRUE'))
                after = self.toks[self.i + 1] if self.i + 1 < len(self.toks) else None
                if after is not None and after.type == 'op' and after.value == '(':
                    return self.parse_call()
                if t.value in RESERVED:
                    fail('E_RESERVED',
                         f'{t.value} is a reserved word and cannot be a variable', t.pos)
                self.next()
                return Node('var', t.pos, name=t.value)

            if t.type == 'op' and t.value == '(':
                self.next()
                if self.at_op(')'):
                    fail('E_SYNTAX', 'empty parentheses', t.pos)
                inner = self.parse_sequence()
                self.expect_op(')')
                # Marked so that F((1,2)) passes one list rather than two arguments.
                inner.grouped = True
                return inner

            fail('E_SYNTAX', f'unexpected {describe(t)}', t.pos)
        finally:
            self.leave()

    def parse_call(self) -> Node:
        name_tok = self.next()
        self.expect_op('(')
        if self.at_op(')'):
            self.next()
            args: list[Node] = []
        else:
            inner = self.parse_sequence()
            self.expect_op(')')
            args = inner.items if (inner.t == 'list' and not inner.grouped) else [inner]

        spec = lookup(name_tok.value)
        if spec is None:
            fail('E_UNKNOWN_FUNC', f'unknown function {name_tok.value}', name_tok.pos)
        if len(args) < spec.min or len(args) > spec.max:
            fail('E_ARITY', f'{spec.name} takes {arity_text(spec)}, got {len(args)}',
                 name_tok.pos)
        if spec.arity_error is not None:
            problem = spec.arity_error(len(args))
            if problem:
                fail('E_ARITY', problem, name_tok.pos)
        return Node('call', name_tok.pos, name=spec.name, spec=spec, args=args)


def arity_text(spec: Spec) -> str:
    if spec.max == float('inf'):
        return f'at least {spec.min} argument{"" if spec.min == 1 else "s"}'
    if spec.min == spec.max:
        return f'{spec.min} argument{"" if spec.min == 1 else "s"}'
    return f'{spec.min} to {spec.max} arguments'


def describe(t: Token) -> str:
    if t.type == 'eof':
        return 'end of input'
    if t.type == 'text':
        return 'a text literal'
    if t.type == 'num':
        return f'number {t.value}'
    return f'"{t.value}"'


def check_target(node: Node, op_tok: Token) -> None:
    """The target must be an identifier followed by zero or more index operations."""
    n = node
    while n.t == 'index':
        n = n.obj
    if n.t != 'var' or node.grouped:
        fail('E_BAD_ASSIGN', f'cannot assign with {op_tok.value} to this expression',
             node.pos)


def parse(source: str) -> Node:
    return Parser(tokenize(source)).parse_program()
